graveler/uncommitted_diff_iterator.go

```diff
@@ -0,0 +1,96 @@
+package graveler
+
+import (
+	"context"
+	"errors"
+
+	"github.com/treeverse/lakefs/logging"
+)
+
+type uncommittedDiffIterator struct {
+	committedManager CommittedManager
+	list             ValueIterator
+	sn               StorageNamespace
+	treeID           TreeID
+	value            *Diff
+	err              error
+}
+
+func NewUncommittedDiffIterator(manager CommittedManager, list ValueIterator, sn StorageNamespace, treeItreeID TreeID) DiffIterator {
+	return &uncommittedDiffIterator{
+		committedManager: manager,
+		list:             list,
+		sn:               sn,
+		treeID:           treeItreeID,
+	}
+}
+
+func valueExistsInCommitted(ctx context.Context, committedManager CommittedManager, sn StorageNamespace, treeID TreeID, key Key) (bool, error) {
+	_, err := committedManager.Get(ctx, sn, treeID, key)
+	if errors.Is(err, ErrNotFound) {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func getDiffType(ctx context.Context, committedManager CommittedManager, sn StorageNamespace, treeID TreeID, key Key, tombstone bool) (DiffType, error) {
+	existsInCommitted, err := valueExistsInCommitted(ctx, committedManager, sn, treeID, key)
+	if err != nil {
+		return 0, err
+	}
+	var diffType DiffType
+	diffType = DiffTypeAdded
+	if tombstone && existsInCommitted {
+		diffType = DiffTypeRemoved
+	}
+	if tombstone && !existsInCommitted {
+		logging.Default().WithFields(logging.Fields{"treeID": treeID, "stagingToken": sn, "key": key}).Warn("tombstone for file that does not exist")
+		diffType = DiffTypeRemoved
+	}
+	if !tombstone && existsInCommitted {
+		diffType = DiffTypeChanged
+	}
+	if !tombstone && !existsInCommitted {
+		diffType = DiffTypeAdded
+	}
+	return diffType, nil
```

Comment:
Think you don't want to continue to fall between all the IFs - so use switch or return at the right place.

```suggestion
	if tombstone {
		if !existsInCommitted {
			logging.Default().
				WithFields(logging.Fields{"tree_id": treeID, "storage_namespace": sn, "key": key}).
				Warn("tombstone for a file that does not exist")
		}
		return DiffTypeRemoved
	}
	if existsInCommitted {
		return DiffTypeChanged
	}
	return DiffTypeAdded
```